doc/developer/design/20210714_reclocking.md

```diff
@@ -0,0 +1,101 @@
+
+Many of our sources of data come with their own understanding of forward progress.
+Often, these are incompatible with each other, and with native notions of time within Materialize.
+For example,
+    Kafka topics contain multiple "parts", each of which advances through "offsets".
+    Postgres sources come from a log where entries have a "log sequence number" (LSN).
+    File sources have a line number.
+    Materialize CDC sources have explicit timestamps.
+
+The source-specific "gauges" of forward progress are useful, but they are not directly comparable.
+This document describes one framework for converting between gauges including in to and potentially out of Materialize's system timeline.
+
+## Goals
+
+The primary goal of this document is to describe a formal way to translate collections that evolve with one gauge into collections that evolve with another gauge.
+That formalism should have the properties that:
+1. There is an explicit representation of the translation (rather than a behavioral description).
+2. The translation should be sidecar metadata, rather than a rewriting of the source data.
+3. The combination of durable source data and durable metadata should result in a durable reclocked stream.
+4. The sidecar metadata should be compactable whenever the reclocked stream is compactable.
+
+The formalism should support the goals of
+A. Permitting the use of streams from one timeline in another timeline.
+B. Describing durable state that can make sources like Kafka, Postgres, and files exactly replayable in the system timeline.
+C. A BYO format for users who want to provide explicit timestamp information for sources that are presented not in that timeline.
+D. A way to describe in one timeline progress made through another timeline (e.g. reporting Kafka read progress in MZ system timestamps).
+
+## Non-Goals
+
+This document has no opinions on the physical representation of the reclocking data.
+This document has no opinions on the implementation of the reclocking actions.
+This document has no opinions on the policies of which streams should be reclocked into which domains.
+This document has no opinions on the correct interpretation of the terms `VOLATILE` and `NONVOLATILE`.
+
+## Description
+
+Streams of data often provide a gauge of progress: stamping records with values and indicating which values will not be seen again.
+This information describes "progress" through a stream, and allows consumers of the stream to rely on the presence of certain contents.
+Multiple records may share the same gauge value, indicating that they should occur at the same moment.
+Records may not be in order of the gauge, but once a value is indicated to have passed it should not appear on a subsequent record.
+
+### `remap`
+
+We are going to use a `remap` collection, which is a map from a target gauge `IntoTime` to an antichain of timestamps from the source gauge `FromTime`.
```

Comment:
antichain?

Reply:
I don't know what you are asking. To replace `timestamps` with `values`, which I should probably do?

Reply:
You changed the type such that it's no longer antichain, but you still say "antichain" in the prose.

Reply:
I see. It should still be an antichain, though the representation is as a collection. But, it has to be a collection that is an antichain.